hls4ml/templates/vivado/nnet_utils/nnet_merge_stream.h

```diff
@@ -299,6 +299,83 @@ void concatenate3d(
     }
 }
 
+template<class input1_T, class input2_T, class res_T, typename CONFIG_T>
+void concatenate2d_0(
+    hls::stream<input1_T> &data1,
+    hls::stream<input2_T> &data2,
+    hls::stream<res_T> &res)
```

Comment:
Unlike the transpose, it makes some sense that these types can be different because `data1` and `data2` could have gone through different layers and acquired different types. That said, there's still a risk that we could run into the kind of issue from #363. Maybe we can do something to help, like set a default value of the output type that would be lossless - like with integer bits `max(data1 integer bits, data2 integer bits)` and `fractional bits max(data1 fractional bits, data2 fractional bits)`

Reply:
Right, I agree. Where do you think we could do this? In the python frontend, I suppose?

Reply:
Yeah in the Python, I think it would make sense to write an optimizer pass to do it

Reply:
@thesps I added an optimizer that attempts to do what you suggest. Namely, it takes the max of the incoming widths
```
newwidth = max(itype1.width, itype2.width)
newint = max(itype1.integer, itype2.integer)
```
but it also tries to account for signed vs unsigned (i.e. add 1 bit if we're mixing signed + unsigned) and if one of the two has a saturation/rounding mode, it takes that (with a preference for input1 if they both have different modes).

I also augmented the test so it triggers the optimizer.

Let me know if you have any suggestions!